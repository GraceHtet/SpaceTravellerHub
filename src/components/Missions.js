import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions } from '../redux/missions/missionsSlice';
import styles from '../styles/Missions.module.css';

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);

  useEffect(() => {
    if (!missions.length) {
      dispatch(fetchMissions());
    }
  }, [dispatch, missions]);

  return (
    <div>
      <h1 className={styles.title}>Missions</h1>

      <table>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <tr key={mission.mission_id}>
              <td>{mission.mission_name}</td>
              <td>{mission.description}</td>
              <td>{mission.status}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};
export default Missions;
