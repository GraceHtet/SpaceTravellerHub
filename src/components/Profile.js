import { useSelector } from 'react-redux';
import styles from '../styles/Profile.module.css';

const Profile = () => {
  const { rocketLists } = useSelector((state) => state.rockets);

  const reserveRockets = rocketLists.filter((rocket) => rocket.reserved && rocket);

  return (
    <div className={styles.container}>
      <div />
      <div className={styles.rocket}>
        <h2>My Rockets</h2>
        <table className={styles.table}>
          <tbody className={styles.tbody}>
            {reserveRockets.map((rocket) => (
              <tr key={rocket.id} className={styles.tr}>
                <td>{rocket.name}</td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};
export default Profile;
