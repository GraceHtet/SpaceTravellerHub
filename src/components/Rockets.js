import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
// eslint-disable-next-line object-curly-newline
import { fetchRockets, reserveRocket } from '../redux/rockets/rocketsSlice';
import styles from '../styles/Rockets.module.css';

const Rockets = () => {
  const dispatch = useDispatch();
  const { rockets } = useSelector((state) => state.rockets);
  const { status } = useSelector((state) => state.rockets);

  useEffect(() => {
    if (status === false) dispatch(fetchRockets());
  }, [dispatch, status]);

  return (
    <div className={styles.container}>
      {status && rockets.map((each) => (
        <article key={each.id} className={styles.article}>
          <img src={each.flickr_images} className={styles.img} alt="rocket" />
          <div>
            <h2 className={styles.title}>{each.name}</h2>
            <p>
              {each.reserved ? <span className={styles.span}>Reserved</span> : ''}
              {each.description}
            </p>
            <button
              type="button"
              className={each.reserved ? styles.unreserve : styles.reserve}
              onClick={() => dispatch(reserveRocket(each.id))}
            >
              {each.reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
            </button>
          </div>
        </article>
      ))}
    </div>
  );
};
export default Rockets;
