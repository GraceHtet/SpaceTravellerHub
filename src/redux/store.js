import { confugureStore } from '@reduxjs/toolkit';

const store = configureStore({
  reducer: {},
});

export default store;
