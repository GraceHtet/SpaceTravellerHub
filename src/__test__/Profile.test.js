import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom/extend-expect';
import TestRenderer, { act } from 'react-test-renderer';
import Rockets from '../components/Rockets';
import Missions from '../components/Missions';
import Profile from '../components/Profile';
import store from '../redux/store';

describe('Missions', () => {
  test('Render Missions Component', () => {
    const tree = TestRenderer.create(
      <Provider store={store}>
        <Missions />
        <Rockets />
        <Profile />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Rocket Button and Misson Button', async () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
      </Provider>,
    );

    const reserveBtn = await screen.findAllByText('Reserve Rocket');
    const joinButton = await screen.findAllByText('Join Mission');
    expect(joinButton).toHaveLength(10);
    expect(reserveBtn).toHaveLength(4);
  });

  test('Reserved Rocket name should be in profile', async () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
        <Profile />
      </Provider>,
    );

    act(() => {
      const reserveBtn = screen.getAllByText('Reserve Rocket');
      reserveBtn[0].click();
    });

    const rocketName = await screen.getAllByText('Falcon 1');
    expect(rocketName[1]).toBeInTheDocument();
  });

  test('Joined Mission name should be in profile', async () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
        <Profile />
      </Provider>,
    );

    act(() => {
      const joinButton = screen.getAllByText('Join Mission');
      joinButton[0].click();
    });

    const missionName = await screen.getAllByText('Thaicom');
    expect(missionName[1]).toBeInTheDocument();
  });
});
