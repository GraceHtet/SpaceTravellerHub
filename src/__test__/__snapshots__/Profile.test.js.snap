// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Render Missions Component 1`] = `
Array [
  <div>
    <h1
      className="title"
    >
      Missions
    </h1>
    <p
      className="loading"
    >
      Loading missions...
    </p>
    <table
      className="table"
    >
      <thead
        className="thead"
      >
        <tr
          className="tr"
        >
          <th
            className="th"
          >
            Mission
          </th>
          <th
            className="th"
          >
            Description
          </th>
          <th
            className="th status"
          >
            Status
          </th>
          <th
            className="th action"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>,
  <div
    className="container"
  />,
  <div
    className="container"
  >
    <div
      className="mission"
    >
      <h2>
        My Missions
      </h2>
      <table
        className="table"
      >
        <tbody
          className="tbody"
        />
      </table>
    </div>
    <div
      className="rocket"
    >
      <h2>
        My Rockets
      </h2>
      <table
        className="table"
      >
        <tbody
          className="tbody"
        />
      </table>
    </div>
  </div>,
]
`;
